designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,406 @@
+- Repo: <https://github.com/eslint/eslint>
+- Start Date: 2024-10-22
+- RFC PR:
+- Authors: 
+  - [Samuel Therrien](https://github.com/Samuel-Therrien-Beslogic) (aka [@Avasam](https://github.com/Avasam))
+  - [Maxim Morev](https://github.com/MorevM)
+
+# Per-rule autofix configuration
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+This RFC proposes adding support for controlling autofix behavior on a per-rule basis through shareable configuration.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+Some rules support autofixing, which is often convenient, but in certain cases 
+the fixes may be broken, unsafe, or simply undesirable. \
+Ideally, unsafe autofixes should be treated as suggestions, and broken fixes should be reported. 
+
+However, ESLint is a large ecosystem, and some useful plugins are no longer actively maintained. 
+Even in actively maintained projects, users may want to disable autofixing for specific rules
+due to project-specific or personal preferences.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+### Abstract
+
+We introduce an *extended rule configuration format*:
+
+```ts
+type RuleEntry = Partial<{
+  /**
+   * The rule severity.
+   * 
+   * @default 2
+   */
+  severity: 0 | 1 | 2 | 'off' | 'warn' | 'error';
+
+  /**
+   * Array of options passed to the rule.
+   * 
+   * @default []
+   */
+  options: any[]
+
+  /**
+   * Whether to allow a rule to perform autofixes.
+   * 
+   * @default true
+   */
+  autofix: boolean;
+}>;
+```
+
+This format enables explicit control over a rule's autofix behavior,
+and opens the door to further extension with other meta-properties
+([example](https://github.com/eslint/eslint/issues/19342)).
+
+> [!IMPORTANT]
+> This is an additional configuration format; existing configurations remain valid and do not need to be rewritten.
+
+#### Autofix details
+
+* The `autofix` option defaults to `true`, maintaining current behavior 
+  (if a rule provides a fixer, it will run when `--fix` is used);
+* If autofix is disabled for a rule, its fix becomes a suggestion labeled 
+  "Apply disabled autofix" and can be applied manually through editor hints;
+* If the rule does not define a fixer, the `autofix` key has no effect.
+
+#### Additional extensions
+
+We extend `linterOptions.reportUnusedDisableDirectives` 
```

Comment:
How does extending these two options relate to allowing users to diasble autofux at the rule level?  This seems like a separate concern, or at least it isn't quite clear to me.

Reply:
I'll agree that it conflicts with the RFC headline.

There was a [discussion thread in the original RFC](https://github.com/eslint/rfcs/pull/125#discussion_r1924282197) and the task seems related to me - I'd rather change the RFC title to something like "Control over autofix changes" than put the change in a separate RFC, as the steps would be quite similar.
Although it would be easier to focus, of course, even though the tasks are related.

Should I remove this suggestion now, or wait for other opinions?

Reply:
I think it's fine to leave it here given that these options can also produce autofixes that people might want to control.